File: legacy/tools/csscheck.py
#!/usr/bin/env python3
"""Every custom class must have a rule, and every rule must be used.

The behavioural suites cannot see the UI layer at all, so a class with no
styling renders as an unstyled element and a rule with no class is dead weight
that will confuse the next person to read it. This caught `sb-swatch`, which
was being styled by an element selector (`.sb-key i`) while the markup named a
class — working by accident, and silently lost the moment either moved.

Reads the built HTML rather than the source so it sees exactly what ships.
"""
import re, sys

path = sys.argv[1] if len(sys.argv) > 1 else 'build/reef-console.html'
html = open(path, encoding='utf-8').read()
app = html.split('<script id="app-src" type="text/plain">')[1].split('</script>')[0]

used = set()
for m in re.finditer(r'className="([^"{]*)"', app):
    used.update(m.group(1).split())
# Template literals, including the branches of a ternary inside them — those
# class names are just as real as the static ones.
for m in re.finditer(r'className=\{`([^`]*)`\}', app):
    body = m.group(1)
    for lit in re.findall(r'"([^"]*)"', body):
        used.update(lit.split())
    for t in re.sub(r'\$\{[^}]*\}', ' ', body).split():
        used.add(t)
for m in re.finditer(r'className=\{[^}]*\?[^}]*\}', app):
    for lit in re.findall(r'"([^"]*)"', m.group(0)):
        used.update(lit.split())

# A fixed prefix list meant a new family was simply invisible: the splash
# classes for the correction celebration had no rules at all and this reported
# OK, because "splash" was not on the list. The checker is only worth having if
# adding a class cannot silently opt out of it — so treat anything that is not
# a known Tailwind utility as custom, and let the allowlist below carry the
# handful of exceptions.
TAILWIND = re.compile(
    r'^(?:'
    r'[a-z]+:'                     # responsive / state prefixes
    r'|-?(?:m|p)[trblxy]?-'
    r'|(?:w|h|min-w|min-h|max-w|max-h|inset|top|right|bottom|left|z)-'
    r'|(?:flex|grid|col|row|gap|order|basis|grow|shrink)(?:-|$)'
    r'|(?:items|justify|content|self|place)-'
    r'|(?:text|font|leading|tracking|whitespace|break|truncate|tabular)(?:-|$)'
    r'|(?:bg|border|rounded|ring|shadow|opacity|outline|divide)(?:-|$)'
    r'|(?:absolute|relative|fixed|sticky|static|block|inline|hidden|table|contents)$'
    r'|(?:overflow|object|cursor|pointer-events|select|resize|appearance)(?:-|$)'
    r'|(?:transition|duration|ease|delay|animate|transform|scale|rotate|translate|origin)(?:-|$)'
    r'|(?:space|backdrop|filter|blur|list|align|uppercase|lowercase|capitalize|italic|underline)(?:-|$)'
    r'|(?:sr-only|not-sr-only|antialiased|shrink|grow|isolate|mix-blend)(?:-|$)'
    r'|(?:aspect|inline-block|inline-flex|inline-grid|float|clear|box|container)(?:-|$)'
    r')'
)
# Utility-looking names that are genuinely ours would go here.
NOT_CUSTOM = set()
custom = sorted(c for c in used if not TAILWIND.match(c) and c not in NOT_CUSTOM)
missing = [c for c in custom if '.' + c not in html]

defined = set(re.findall(r'\.((?:brief|strip|sb|el|hero|param-row|session)[a-z0-9-]*)\s*[,{:]', html))
unused = sorted(d for d in defined if d not in used)

problems = []
for c in missing:
    problems.append(f'class used with no CSS rule: {c}')
for d in unused:
    problems.append(f'CSS rule never used: {d}')

if problems:
    print('FAIL stylesheet and markup disagree:')
    for p in problems:
        print(f'    {p}')
    sys.exit(1)
print(f'OK   stylesheet matches markup ({len(custom)} custom classes)')

File: scripts/verify/csscheck.mjs
#!/usr/bin/env node
/* Ported from legacy/tools/csscheck.py, with changes — routine flagged this
 * as the checker most likely not to port, and half of it isn't: the "every
 * Tailwind utility has a rule" question does not exist here, because
 * Tailwind generates rules for the utility classes it recognises on demand
 * and purges the rest at build time — there is no fixed stylesheet where a
 * utility class could go unstyled.
 *
 * What Tailwind's build does NOT check is the other half: hand-authored
 * classes (`brief-*`, `sb-*`, `rc-*`, `sp-*`, ...) are still real, still
 * live in src/styles/*.css, not generated, not purged, and a typo in either
 * direction — a className with no rule, a rule nothing renders — ships
 * silently. That's the same fault csscheck.py was written for (it caught
 * `sb-swatch` styled by an element selector while the markup named a class,
 * working by accident). Reading source directly instead of built HTML is a
 * strict improvement, not a cost: no build step is needed to catch a typo
 * introduced in either file before it ever compiles.
 *
 * Also widened from the monolith's fixed custom-class prefix list
 * (`brief|strip|sb|el|hero|param-row|session`) to "not a recognised Tailwind
 * pattern" for BOTH directions — the original only used that exclusion for
 * classes found in use; the unused-rule side still filtered through the
 * fixed prefix list, which is the exact "whole class family invisible"
 * failure mode the docstring describes, just left unfixed on one side.
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT, listSrcFiles, readAll } from './util.mjs';

const TAILWIND = new RegExp(
  '^(?:' +
  '[a-z-]+:' +                    // responsive / state prefixes
  '|-?(?:m|p)[trblxy]?-' +
  '|(?:w|h|min-w|min-h|max-w|max-h|inset|top|right|bottom|left|z)-' +
  '|(?:flex|grid|col|row|gap|order|basis|grow|shrink)(?:-|$)' +
  '|(?:items|justify|content|self|place)-' +
  '|(?:text|font|leading|tracking|whitespace|break|truncate|tabular)(?:-|$)' +
  '|(?:bg|border|rounded|ring|shadow|opacity|outline|divide)(?:-|$)' +
  '|(?:absolute|relative|fixed|sticky|static|block|inline|hidden|table|contents)$' +
  '|(?:overflow|object|cursor|pointer-events|select|resize|appearance)(?:-|$)' +
  '|(?:transition|duration|ease|delay|animate|transform|scale|rotate|translate|origin)(?:-|$)' +
  '|(?:space|backdrop|filter|blur|list|align|uppercase|lowercase|capitalize|italic|underline)(?:-|$)' +
  '|(?:sr-only|not-sr-only|antialiased|shrink|grow|isolate|mix-blend)(?:-|$)' +
  '|(?:aspect|inline-block|inline-flex|inline-grid|float|clear|box|container)(?:-|$)' +
  ')'
);

const files = listSrcFiles();
const entries = readAll(files.filter((f) => f.endsWith('.jsx')));

const used = new Set();
for (const { src } of entries) {
  for (const m of src.matchAll(/className="([^"{]*)"/g)) m[1].split(/\s+/).forEach((c) => c && used.add(c));
  for (const m of src.matchAll(/className=\{`([^`]*)`\}/g)) {
    const bodyStr = m[1];
    // Only strings in ternary-branch position (`? "x" : "y"`) count as class
    // literals — an unqualified `"([^"]*)"` scan also catches a plain string
    // comparison inside the same expression (`status === "ok" ? ... : ...`)
    // and reads its value as a class name.
    for (const lit of bodyStr.matchAll(/[?:]\s*"([^"]*)"/g)) lit[1].split(/\s+/).forEach((c) => c && used.add(c));
    bodyStr.replace(/\$\{[^}]*\}/g, ' ').split(/\s+/).forEach((c) => c && used.add(c));
  }
  for (const m of src.matchAll(/className=\{([^`}]*\?[^}]*)\}/g)) {
    for (const lit of m[1].matchAll(/[?:]\s*"([^"]*)"/g)) lit[1].split(/\s+/).forEach((c) => c && used.add(c));
  }
}

const cssDir = path.join(ROOT, 'src', 'styles');
const cssFiles = fs.readdirSync(cssDir).filter((f) => f.endsWith('.css')).map((f) => path.join(cssDir, f));
const css = cssFiles.map((f) => fs.readFileSync(f, 'utf8')).join('\n');

const custom = [...used].filter((c) => !TAILWIND.test(c)).sort();
const missing = custom.filter((c) => !css.includes('.' + c));

// Every hand-authored selector (i.e. not a Tailwind pattern) declared in the
// stylesheet — both sides of the comparison now use the same exclusion rule.
const declaredSelectors = new Set([...css.matchAll(/\.([a-zA-Z][\w-]*)\s*[,{:]/g)].map((m) => m[1]));
const defined = [...declaredSelectors].filter((d) => !TAILWIND.test(d)).sort();
const unused = defined.filter((d) => !used.has(d));

const problems = [];
for (const c of missing) problems.push(`class used with no CSS rule: ${c}`);
for (const d of unused) problems.push(`CSS rule never used: ${d}`);

if (problems.length) {
  console.log('FAIL stylesheet and markup disagree:');
  for (const p of problems) console.log(`    ${p}`);
  process.exit(1);
}
console.log(`OK   stylesheet matches markup (${custom.length} custom classes, ${cssFiles.length} CSS files)`);
